File: Trabalho 1/distance.py
"""
Requisito 5 do enunciado: distancia entre dois vertices e diametro do grafo.

A distancia usa a BFS como primitiva, como pede o enunciado: em grafo nao
ponderado, o nivel da BFS E a distancia minima.

Diametro exato: BFS a partir de cada vertice, O(n * (n + m)). Inviavel nos
grafos grandes do estudo de caso, por isso ha tambem a versao aproximada.

Diametro aproximado (heuristica de dupla varredura): a partir de um vertice
qualquer, a BFS acha o vertice mais distante u; uma segunda BFS a partir de u
da uma estimativa por baixo do diametro. E barato (2 BFS por componente) e
costuma acertar ou chegar muito perto em grafos reais.
"""

from my_queue import Queue
from graph_utils import neighbors

INFINITO = -1  # marcador de "inalcancavel", coerente com levels = -1 do BFS.py


def bfs_distances(graph, source, adj_list: bool = False):
    """
    Distancias de `source` (0-based) para todos os vertices.
    Retorna lista onde nao alcancado = INFINITO.
    """
    n = len(graph)
    distances = [INFINITO] * n
    distances[source] = 0

    queue = Queue()
    queue.enqueue(source)

    while not queue.isEmpty():
        current = queue.dequeue()
        for neighbor in neighbors(graph, current, adj_list):
            if distances[neighbor] == INFINITO:
                distances[neighbor] = distances[current] + 1
                queue.enqueue(neighbor)

    return distances


def distance(graph, u, v, adj_list: bool = False, one_based: bool = True):
    """
    Distancia entre u e v. Por padrao recebe rotulos 1-based (como no arquivo
    de entrada) e devolve INFINITO se estiverem em componentes diferentes.
    """
    if one_based:
        u, v = u - 1, v - 1
    return bfs_distances(graph, u, adj_list)[v]


def _eccentricity(graph, source, adj_list):
    """Maior distancia finita a partir de source, e o vertice que a realiza."""
    distances = bfs_distances(graph, source, adj_list)
    best_vertex, best_distance = source, 0

    for vertex, d in enumerate(distances):
        if d != INFINITO and d > best_distance:
            best_vertex, best_distance = vertex, d

    return best_distance, best_vertex


def diameter(graph, adj_list: bool = False):
    """
    Diametro exato: maior distancia minima entre qualquer par alcancavel.
    Em grafo desconexo, e o maior diametro entre as componentes (distancias
    infinitas sao ignoradas).
    """
    largest = 0
    for source in range(len(graph)):
        eccentricity, _ = _eccentricity(graph, source, adj_list)
        if eccentricity > largest:
            largest = eccentricity
    return largest


def approximate_diameter(graph, adj_list: bool = False):
    """
    Estimativa por dupla varredura, aplicada uma vez por componente.
    O valor devolvido e sempre <= diametro real (e uma cota inferior).
    """
    from components import connected_components

    estimate = 0
    for component in connected_components(graph, adj_list):
        start = component[0] - 1  # rotulo 1-based -> indice

        _, farthest = _eccentricity(graph, start, adj_list)
        second_pass, _ = _eccentricity(graph, farthest, adj_list)

        if second_pass > estimate:
            estimate = second_pass

    return estimate


if __name__ == "__main__":
    from pathlib import Path
    from graph import Graph

    example = Path(__file__).with_name("graph_exemplo.txt")

    for use_list in (False, True):
        rep = "lista de adjacencia" if use_list else "matriz de adjacencia"
        g = Graph(example, adj_list=use_list)
        print(f"--- {rep} ---")
        print("d(1,2) =", distance(g.graph, 1, 2, use_list))
        print("d(1,3) =", distance(g.graph, 1, 3, use_list))
        print("d(3,4) =", distance(g.graph, 3, 4, use_list))
        print("diametro exato       =", diameter(g.graph, use_list))
        print("diametro aproximado  =", approximate_diameter(g.graph, use_list))
        print()

File: Trabalho 1 (Implementação em C)/Algorithms/distance.c
#include <stdlib.h>

#include "distance.h"
#include "components.h"
#include "my_queue.h"
#include "graph_utils.h"

int *bfs_distances(Graph *g, int source) {
    /*
    Distancias de `source` (0-based) para todos os vertices.
    Retorna vetor onde nao alcancado = INFINITO.
    */
    int n = g->n;
    int *distances = xmalloc((size_t)n * sizeof(int));
    for (int v = 0; v < n; v++) {
        distances[v] = INFINITO;
    }
    distances[source] = 0;

    Queue queue;
    Queue_init(&queue);
    Queue_enqueue(&queue, source);

    while (!Queue_isEmpty(&queue)) {
        int current = Queue_dequeue(&queue);

        int neighbor_count;
        int *neighbor = neighbors(g, current, &neighbor_count);
        for (int j = 0; j < neighbor_count; j++) {
            if (distances[neighbor[j]] == INFINITO) {
                distances[neighbor[j]] = distances[current] + 1;
                Queue_enqueue(&queue, neighbor[j]);
            }
        }
        free(neighbor);
    }

    return distances;
}

int distance(Graph *g, int u, int v, bool one_based) {
    /*
    Distancia entre u e v. Por padrao recebe rotulos 1-based (como no arquivo
    de entrada) e devolve INFINITO se estiverem em componentes diferentes.
    */
    if (one_based) {
        u -= 1;
        v -= 1;
    }

    int *distances = bfs_distances(g, u);
    int result = distances[v];
    free(distances);
    return result;
}

/* _eccentricity no Python: privada do modulo (static em C) */
static void eccentricity(Graph *g, int source, int *best_distance_out, int *best_vertex_out) {
    /* Maior distancia finita a partir de source, e o vertice que a realiza. */
    int *distances = bfs_distances(g, source);
    int best_vertex = source;
    int best_distance = 0;

    for (int vertex = 0; vertex < g->n; vertex++) {
        if (distances[vertex] != INFINITO && distances[vertex] > best_distance) {
            best_vertex = vertex;
            best_distance = distances[vertex];
        }
    }

    free(distances);
    *best_distance_out = best_distance;
    *best_vertex_out = best_vertex;
}

int diameter(Graph *g) {
    /*
    Diametro exato: maior distancia minima entre qualquer par alcancavel.
    Em grafo desconexo, e o maior diametro entre as componentes (distancias
    infinitas sao ignoradas).
    */
    int largest = 0;
    for (int source = 0; source < g->n; source++) {
        int eccentricity_value;
        int vertex;
        eccentricity(g, source, &eccentricity_value, &vertex);
        if (eccentricity_value > largest) {
            largest = eccentricity_value;
        }
    }
    return largest;
}

int approximate_diameter(Graph *g) {
    /*
    Estimativa por dupla varredura, aplicada uma vez por componente.
    O valor devolvido e sempre <= diametro real (e uma cota inferior).
    */
    Components components = connected_components(g);

    int estimate = 0;
    for (int i = 0; i < components.count; i++) {
        int start = components.components[i][0] - 1; /* rotulo 1-based -> indice */

        int primeira;
        int farthest;
        eccentricity(g, start, &primeira, &farthest);

        int second_pass;
        int vertex;
        eccentricity(g, farthest, &second_pass, &vertex);

        if (second_pass > estimate) {
            estimate = second_pass;
        }
    }

    Components_free(&components);
    return estimate;
}

#ifdef DISTANCE_MAIN
/* if __name__ == "__main__" do distance.py */
#include <stdio.h>

int main(void) {
    for (int use_list = 0; use_list <= 1; use_list++) {
        const char *rep = use_list ? "lista de adjacencia" : "matriz de adjacencia";
        Graph *g = Graph_create("graph_exemplo.txt", use_list != 0);
        printf("--- %s ---\n", rep);
        printf("d(1,2) = %d\n", distance(g, 1, 2, true));
        printf("d(1,3) = %d\n", distance(g, 1, 3, true));
        printf("d(3,4) = %d\n", distance(g, 3, 4, true));
        printf("diametro exato       = %d\n", diameter(g));
        printf("diametro aproximado  = %d\n", approximate_diameter(g));
        printf("\n");
        Graph_destroy(g);
    }
    return 0;
}
#endif
